pysparkling/rdd.py

```diff
@@ -849,6 +888,64 @@ def leftOuterJoin(self, other, numPartitions=None):
             for v_other in (d_other[kv[0]] if kv[0] in d_other else [None])
         ])
 
+    def _leftSemiJoin(self, other):
```

Comment:
The leading underscore means "private" in Python and the linter will want to enforce that. 
What about an underscore at the end of the name?

Reply:
I believe it is better to make this method private:
It would be annoying for a user to let him believe it's available in Spark and let him build a working pipeline until he tries with Pyspark and he realize he needs to redesign his whole flow,

I even believe that autocompletion would even encourage him to use this method as it is not private,

It is only called once in the Spark SQL code (in `DataFrame` `join` method), so adding a lint seems ok to me

What do you think?

Reply:
@svenkreiss: If this is the last point blocking the merge I can remove those methods and we can talk about them later on, they can be extracted into functions and put where Spark SQL needs them, outside of this class, which would remove both our blocking points (no linter error and they are clearly outside of the Spark API)

Reply:
Sorry, I was super busy.

I think the question of Spark compatibility should not interfere with OO design. In some prior issues somewhere, there was already a discussion to introduce a `spark_compatible` flag to `RDD` or `context` such that a function like the one you introduced can raise an Exception if it is called and the flag is set.

There will be other functions probably that we will want to introduce that are not in PySpark but generally useful and we need a mechanism for that.

Your thoughts on this are very welcome.

Reply:
I don't think it would be easy to pass reference to this flag to all the function of the library,

A solution would be to log a warning whenever some flag is set in an environment variable,

It is imperfect but make it easy to implement,

I do not know what to do when some implementation details of pysparkling are used though (e.g. RDD.sort is an in order sort, whereas Spark's one is not)

Reply:
I don't think it would be easy to pass reference to this flag to all the function of the library,

A solution would be to log a warning whenever some flag is set in an environment variable,

It is imperfect but make it easy to implement,

I do not know what to do when some implementation details of pysparkling are used though (e.g. RDD.sort is an in order sort, whereas Spark's one is not)

Reply:
There are other options but we don't have to solve it now. Let's just revisit this later.